spec/support/thread_order_support.rb

```diff
@@ -0,0 +1,93 @@
+class ThreadOrderSupport
+  def initialize
+    @bodies  = {}
+    @threads = []
+    @queue   = [] # we may not have thread stdlib required, so may not have Queue class
+    @mutex   = Mutex.new
+    @worker  = Thread.new { loop { work } }
+    @worker.abort_on_exception = true
+  end
+
+  def declare(name, &block)
+    @bodies[name] = block
+  end
+
+  def current
+    Thread.current[:thread_order_name]
+  end
+
+  def pass_to(name, options)
+    parent       = Thread.current
+    child        = nil
+    resume_event = extract_resume_event! options
+    resume_if    = lambda do |event|
+      return unless event == sync { resume_event }
+      parent.wakeup
+    end
+
+    enqueue do
+      child = Thread.new do
+        enqueue { @threads << child }
+        sync { resume_event } == :sleep &&
+          enqueue { watch_for_sleep(child) { resume_if.call :sleep } }
+        begin
+          enqueue { resume_if.call :run }
+          Thread.current[:thread_order_name] = name
+          @bodies.fetch(name).call
+        rescue Exception => error
+          enqueue { parent.raise error }
+          raise
+        ensure
+          enqueue { resume_if.call :exit }
+        end
+      end
+    end
+
+    sleep
+    child
+  end
+
+  def apocalypse!(thread_method=:kill)
+    enqueue do
+      @threads.each(&thread_method)
+      @queue.clear
+      @worker.kill
+    end
+    @worker.join
+  end
+
+  private
+
+  def sync(&block)
+    @mutex.synchronize(&block)
+  end
+
+  def enqueue(&block)
+    sync { @queue << block }
+  end
+
+  def work
+    task = sync { @queue.shift }
+    task ||= lambda { Thread.pass }
+    task.call
+  end
+
+  def extract_resume_event!(options)
+    resume_on = options.delete :resume_on
+    options.any? &&
+      raise(ArgumentError, "Unknown options: #{options.inspect}")
+    resume_on && ![:run, :exit, :sleep].include?(resume_on) and
+      raise(ArgumentError, "Unknown status: #{resume_on.inspect}")
+    resume_on
+  end
+
+  def watch_for_sleep(thread, &cb)
+    if thread.status == false || thread.status == nil
+      # noop, dead threads dream no dreams
+    elsif thread.status == 'sleep'
+      cb.call
+    else
+      enqueue { watch_for_sleep(thread, &cb) }
+    end
+  end
+end
```

Comment:
If you're referring to `watch_for_sleep`, it check to see if the thread died (`false`) or errored (`nil`), which means it won't ever sleep, so in this case it returns, which means it won't continue watching. If the status is `"sleep"`, then it has found what it's watching for, so it invokes the callback and returns, which causes it to stop watching. If the thread is not dead/errored/asleep, then it needs to keep watching, so it enqueues another call to `watch_for_sleep`, which means that the worker will eventually check again.

If you're referring to the implementation of the class as a whole, then yeah, it's a bit much, I struggled to understand it, too. It took 12 hours to write the thing, because it took that long to get my brain to really consider the various implications of concurrency. My way of thinking about it is that it's probably a shitty implementation of what `node` does. If there's some good way to delete it, or simplify it, I would love to see that. Right now, I can't really think of any, and I'm only 80-90% confident that it *actually* works (I test it by running it on rbx 20 times in a row without deadlocks or failures).